// tello-drone/lib/drone.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const dgram_1 = __importDefault(require("dgram"));
const events_1 = require("events");
const utility_1 = require("./utility");
class Drone {
    constructor(options) {
        /*
            Leaving these asserts out for now, since changing the ports from string to number
            as per following the dgram docs, it will cause errors. Right now the goal is to port JS to TS
            and not fix bugs. Will be added in 3.1.x release.
        */
        // assert.equal(typeof options.host, "string");
        // assert.equal(typeof options.port, "number");
        // assert.equal(typeof options.statePort, "number");
        // assert.equal(typeof options.skipOk, "boolean");
        this.on = (event, callback) => {
            this.events.on(event, callback);
        };
        const defaultOptions = {
            host: "192.168.10.1",
            port: 8889,
            statePort: 8890,
            skipOk: true,
        };
        const { host, port, statePort, skipOk, } = Object.assign(Object.assign({}, defaultOptions), options);
        this.HOST = host;
        this.MAIN_PORT = port;
        this.STATE_PORT = statePort;
        this.droneIO = dgram_1.default.createSocket("udp4");
        this.droneState = dgram_1.default.createSocket("udp4");
        this.droneIO.bind(this.MAIN_PORT);
        this.droneState.bind(this.STATE_PORT);
        this.connected = false;
        this.events = new events_1.EventEmitter();
        this.droneState.on("message", stateBuffer => {
            this.events.emit("state", utility_1.parseDroneState(stateBuffer));
        });
        this.droneIO.on("message", (...args) => {
            const [messageBuffer] = args;
            const parsedMessage = Buffer.isBuffer(messageBuffer) ? messageBuffer.toString() : messageBuffer;
            if (parsedMessage !== "ok") {
                return this.events.emit("message", parsedMessage);
            }
            this.events.emit("_ok");
            if (!this.connected) {
                this.connected = true;
                this.events.emit("connection");
            }
            if (!skipOk)
                this.events.emit("message", parsedMessage);
        });
        // Add a minor delay so that the events can be attached first
        setTimeout(() => this.send("command"));
    }
    send(command, options, force = false) {
        const error = utility_1.verifyCommand(command, options);
        let formattedCommand = command;
        if (options) {
            formattedCommand = utility_1.formatCommand(command, options);
        }
        return new Promise((resolve, reject) => {
            if (error && !force)
                return reject(error);
            this.droneIO.send(formattedCommand, 0, formattedCommand.length, this.MAIN_PORT, this.HOST, this.events.emit.bind(this.events, "send"));
            // I need to really double check this below, but i will not change it for now since im only porting to TS
            // If the command is a read command resolve immediately
            if (formattedCommand.includes("?")) {
                return resolve();
            }
            // otherwise wait for OK from drone
            this.events.once("_ok", resolve);
        });
    }
    forceSend(command, options) {
        return this.send.call(this, command, options, true);
    }
}
module.exports = Drone;
