// @node-tello/commander/dist/enums/ControlCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ControlCommandEnum = void 0;
var ControlCommandEnum;
(function (ControlCommandEnum) {
    ControlCommandEnum["command"] = "command";
    ControlCommandEnum["takeoff"] = "takeoff";
    ControlCommandEnum["land"] = "land";
    ControlCommandEnum["streamon"] = "streamon";
    ControlCommandEnum["streamoff"] = "streamoff";
    ControlCommandEnum["emergency"] = "emergency";
    ControlCommandEnum["up"] = "up";
    ControlCommandEnum["down"] = "down";
    ControlCommandEnum["left"] = "left";
    ControlCommandEnum["right"] = "right";
    ControlCommandEnum["forward"] = "forward";
    ControlCommandEnum["back"] = "back";
    ControlCommandEnum["cw"] = "cw";
    ControlCommandEnum["ccw"] = "ccw";
    ControlCommandEnum["flip"] = "flip";
    ControlCommandEnum["go"] = "go";
    ControlCommandEnum["stop"] = "stop";
    ControlCommandEnum["curve"] = "curve";
    ControlCommandEnum["jump"] = "jump";
})(ControlCommandEnum = exports.ControlCommandEnum || (exports.ControlCommandEnum = {}));
